### PR TITLE
Introduce ManagedBy field in RunPolicy

### PR DESCRIPTION
**Pull request by mszadkow**
Add support for ManagedBy to be able delegate reconciliation process to external controller.

Fixes https://github.com/kubeflow/mpi-operator/issues/646

### CODE REVIEW

File: pkg/apis/kubeflow/v2beta1/types.go [hunk 1]
```diff
@@ -131,6 +139,16 @@ type RunPolicy struct {
 	// Defaults to false.
 	// +kubebuilder:default:=false
 	Suspend *bool `json:"suspend,omitempty"`
+
+	// ManagedBy is used to indicate the controller or entity that manages a job.
```
Comment (tenzen-y) [inline pkg/apis/kubeflow/v2beta1/types.go:143]:
```suggestion
	// ManagedBy is used to indicate the controller or entity that manages a MPIJob.
```

Let's identify which Job is delegated to the external controller.
Because the mpi-operator has batch/v1 Job concept.

File: pkg/apis/kubeflow/v2beta1/types.go [hunk 2]
```diff
@@ -131,6 +139,16 @@ type RunPolicy struct {
...
+	// The value must be either an empty, 'kubeflow.org/mpi-operator' or
+	// 'kueue.x-k8s.io/multikueue'.
+	// The mpi-operator reconciles a job which doesn't have this
```
Comment (tenzen-y) [inline pkg/apis/kubeflow/v2beta1/types.go:146]:
```suggestion
	// The mpi-operator reconciles a MPIJob which doesn't have this
```

File: pkg/apis/kubeflow/v2beta1/types.go [hunk 3]
```diff
@@ -131,6 +139,16 @@ type RunPolicy struct {
...
+	// field at all or the field value is the reserved string
+	// 'kubeflow.org/mpi-operator', but delegates reconciling the job
```
Comment (tenzen-y) [inline pkg/apis/kubeflow/v2beta1/types.go:148]:
```suggestion
	// 'kubeflow.org/mpi-operator', but delegates reconciling the MPIJob
```

File: pkg/controller/mpi_job_controller_test.go [hunk 4]
```diff
@@ -504,6 +504,18 @@ func TestDoNothingWithInvalidMPIJob(t *testing.T) {
...
+	f := newFixture(t, "")
+	var replicas int32 = 1
+	startTime := metav1.Now()
+	completionTime := metav1.Now()
+	mpiJob := newMPIJob("test", &replicas, &startTime, &completionTime)
+	mpiJob.Spec.MPIImplementation = kubeflow.MPIImplementationOpenMPI
+	mpiJob.Spec.RunPolicy.ManagedBy = ptr.To(kubeflow.MultiKueueController)
+	f.setUpMPIJob(mpiJob)
+	f.run(getKey(mpiJob, t))
+}
```
Comment (tenzen-y) [inline pkg/controller/mpi_job_controller_test.go:517]: Could we verify if secret, launcher, and workers here?

File: test/e2e/mpi_job_test.go [hunk 5]
```diff
@@ -164,6 +165,27 @@ var _ = ginkgo.Describe("MPIJob", func() {
 				mpiJob := createJobAndWaitForCompletion(mpiJob)
 				expectConditionToBeTrue(mpiJob, kubeflow.JobSucceeded)
 			})
+
+			ginkgo.It("should not be updated when managed externaly, only created", func() {
+				mpiJob.Spec.RunPolicy.ManagedBy = ptr.To(kubeflow.MultiKueueController)
+				ctx := context.Background()
+				mpiJob = createJob(ctx, mpiJob)
+
+				gomega.Consistently(func() error {
```
Comment (tenzen-y) [inline test/e2e/mpi_job_test.go:174]:
```suggestion
				gomega.Eventually(func() error {
```

I guess that the `Eventually` should be enough since this verification is just validation testing, right?

File: test/e2e/mpi_job_test.go [hunk 6]
```diff
@@ -164,6 +165,27 @@ var _ = ginkgo.Describe("MPIJob", func() {
...
+					mpiJob = updatedJob
+					return nil
+				}, 5*time.Second, waitInterval).Should(gomega.Succeed())
+
+				// job should be created, but status should not be updated neither for create nor for any other status
+				condition := getJobCondition(mpiJob, kubeflow.JobCreated)
+				gomega.Expect(condition).To(gomega.BeNil())
+				condition = getJobCondition(mpiJob, kubeflow.JobSucceeded)
+				gomega.Expect(condition).To(gomega.BeNil())
+			})
```
Comment (tenzen-y) [inline test/e2e/mpi_job_test.go:188]: Could you verify if Secret, Launcher, and Workers have never been created?

File: pkg/controller/mpi_job_controller_test.go [hunk 7]
```diff
@@ -504,6 +504,18 @@ func TestDoNothingWithInvalidMPIJob(t *testing.T) {
...
+
```
Comment (tenzen-y) [inline pkg/controller/mpi_job_controller_test.go:518]: Additionally, could you add if the MPIJobs with `kubeflow.org/mpi-operator` are reconciled by the mpi-operator?

File: test/e2e/mpi_job_test.go (hunk 5 above)
Comment (mszadkow) [reply test/e2e/mpi_job_test.go:174]: eventually assumes something will change over time and we wait for the condition, consistently proves us nothing changed - we want the latter

File: pkg/controller/mpi_job_controller_test.go (hunk 7 above)
Comment (mszadkow) [reply pkg/controller/mpi_job_controller_test.go:518]: I could add the mpi-operator in `newMPIJobCommon`, so all other tests verify that, although they should already as this is default value
```
	if policy.ManagedBy == nil {
		policy.ManagedBy = ptr.To(KubeflowJobController)
	}
```

Comment (mszadkow) [reply pkg/controller/mpi_job_controller_test.go:526]: I figured that proving no kubeActions should be sufficient, wdyt?

File: test/e2e/mpi_job_test.go (hunk 6 above)
Comment (mszadkow) [reply test/e2e/mpi_job_test.go:197]: sure, done

File: pkg/controller/mpi_job_controller_test.go (hunk 4 above)
Comment (mszadkow) [reply pkg/controller/mpi_job_controller_test.go:525]: done

File: pkg/apis/kubeflow/v2beta1/types.go [hunk 8]
```diff
@@ -131,6 +139,16 @@ type RunPolicy struct {
 	// Defaults to false.
 	// +kubebuilder:default:=false
 	Suspend *bool `json:"suspend,omitempty"`
+
+	// ManagedBy is used to indicate the controller or entity that manages a MPIJob.
+	// The value must be either an empty, 'kubeflow.org/mpi-operator' or
+	// 'kueue.x-k8s.io/multikueue'.
```
Comment (alculquicondor) [inline pkg/apis/kubeflow/v2beta1/types.go:145]: This API cannot say that these are the only possible values. This API shouldn't even be aware that Kueue exists.

The only thing that it could say is that if the value is not `kubeflow.org/mpi-operator` then it's not managed by this controller. See https://github.com/kubernetes/api/blob/71385f038c1097af36f3d2f68b415860b866c1f8/batch/v1/types.go#L473 for an example of how it should look like.

File: pkg/apis/kubeflow/validation/validation.go [hunk 9]
```diff
@@ -39,8 +39,11 @@ var (
 
 	validRestartPolicies = sets.NewString(
 		string(kubeflow.RestartPolicyNever),
-		string(kubeflow.RestartPolicyOnFailure),
-	)
+		string(kubeflow.RestartPolicyOnFailure))
+
+	validManagedBy = sets.NewString(
```
Comment (alculquicondor) [inline pkg/apis/kubeflow/validation/validation.go:44]: The validation should be similar to https://github.com/kubernetes/kubernetes/blob/b4ce3a55bc7218a6710ff9340dccd403eee5e293/pkg/apis/batch/validation/validation.go#L220

File: pkg/controller/mpi_job_controller_test.go [hunk 10]
```diff
@@ -504,6 +509,21 @@ func TestDoNothingWithInvalidMPIJob(t *testing.T) {
 	f.run(getKey(mpiJob, t))
 }
 
+func TestDoNothingWithMPIJobManagedExternally(t *testing.T) {
```
Comment (alculquicondor) [inline pkg/controller/mpi_job_controller_test.go:512]: Also validate that a nil value the ManagedBy field leads to processing.

This is important for jobs that were created before the ManagedBy field existed

File: test/e2e/mpi_job_test.go [hunk 11]
```diff
@@ -164,8 +165,43 @@ var _ = ginkgo.Describe("MPIJob", func() {
...
+				mpiJob = createJob(ctx, mpiJob)
+
+				gomega.Consistently(func() error {
+					updatedJob, err := mpiClient.KubeflowV2beta1().MPIJobs(mpiJob.Namespace).Get(ctx, mpiJob.Name, metav1.GetOptions{})
+					if err != nil {
+						return err
+					}
+					mpiJob = updatedJob
+					return nil
+				}, 5*time.Second, waitInterval).Should(gomega.Succeed())
```
Comment (alculquicondor) [inline test/e2e/mpi_job_test.go:181]: this is too much time. Maybe just 2 seconds.

File: pkg/apis/kubeflow/v2beta1/types.go (hunk 8 above)
Comment (tenzen-y) [reply pkg/apis/kubeflow/v2beta1/types.go:145]: Here is background of the reason why we limit the acceptable values in the managedBy resource.
Ideally, we should enhance the status validations similar to batch/Job and allow managedBy field to be specified arbitrary values
so that we can provide the appropriate rail.

But, the proposal was rejected in the training-operator since the TrainigOperator community is working on the v2 API, and they have no enough bandwidth to implement the additional status validation implementation. And, the v2 API will provided the non restricted managedBy feature same as batch/Job.

So, I think we can take either approach: 1. enhance status validation and allow managedBy to specify arbitrary values 2. Limit the acceptable values kueue or mpi-operator.

If anyone can lead the design and handle the status validation enhancement, I'm ok with opt 1.

Comment (alculquicondor) [reply pkg/apis/kubeflow/v2beta1/types.go:145]: I see, that's unfortunate.

Let's just do the same as the training-operator.

For future reference, the status validations refer to the equivalent of this https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/4368-support-managed-by-for-batch-jobs#job-status-validation

Comment (tenzen-y) [reply pkg/apis/kubeflow/v2beta1/types.go:145]: Sorry, I forgot to add the contextual issue 😓 
https://github.com/kubeflow/training-operator/issues/2193

Comment (tenzen-y) [reply pkg/apis/kubeflow/v2beta1/types.go:145]: > I see, that's unfortunate.
> 
> Let's just do the same as the training-operator.
> 
> For future reference, the status validations refer to the equivalent of this https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/4368-support-managed-by-for-batch-jobs#job-status-validation

Yes, ideally, we should do it. Maybe, we can leave the dream on the TrainJob 😓

File: pkg/controller/mpi_job_controller_test.go (hunk 10 above)
Comment (mszadkow) [reply pkg/controller/mpi_job_controller_test.go:512]: Sure, we could add extra test case, but all other tests use mpiJob returned from `newMPIJob`, where managedBy is nil by default.
Although I could enhance that by adding it explicitly:
```go
func newMPIJobCommon(name string, startTime, completionTime *metav1.Time) *kubeflow.MPIJob {
...
	mpiJob := &kubeflow.MPIJob{
...
		Spec: kubeflow.MPIJobSpec{
			RunPolicy: kubeflow.RunPolicy{
				CleanPodPolicy: &cleanPodPolicyAll,
				ManagedBy: nil,
			},
```
wdyt?

File: test/e2e/mpi_job_test.go (hunk 11 above)
Comment (mszadkow) [reply test/e2e/mpi_job_test.go:181]: Yes, sure, it should be sufficient

File: pkg/apis/kubeflow/validation/validation.go (hunk 9 above)
Comment (alculquicondor) [reply pkg/apis/kubeflow/validation/validation.go:44]: There is no point on following my suggestion if the values are fixed. Forget about this for now.

File: pkg/controller/mpi_job_controller_test.go (hunk 10 above)
Comment (alculquicondor) [reply pkg/controller/mpi_job_controller_test.go:512]: No, that's not necessary. Thanks for checking.

**Comment by alculquicondor** [comment]
@tenzen-y anything else to add?

File: test/e2e/mpi_job_test.go (hunk 5 above)
Comment (tenzen-y) [reply test/e2e/mpi_job_test.go:174]: That makes sense. Thanks!

File: test/e2e/mpi_job_test.go [hunk 12]
```diff
@@ -164,8 +165,43 @@ var _ = ginkgo.Describe("MPIJob", func() {
...
+					mpiJob = updatedJob
+					return nil
+				}, 2*time.Second, waitInterval).Should(gomega.Succeed())
+
+				// job should be created, but status should not be updated neither for create nor for any other status
+				condition := getJobCondition(mpiJob, kubeflow.JobCreated)
+				gomega.Expect(condition).To(gomega.BeNil())
+				condition = getJobCondition(mpiJob, kubeflow.JobSucceeded)
+				gomega.Expect(condition).To(gomega.BeNil())
+				launcherPods, err := getLauncherPods(ctx, mpiJob)
```
Comment (tenzen-y) [inline test/e2e/mpi_job_test.go:188]: Could you verify if the Launcher Job doesn't exist instead of the Launcher Pod?

File: test/integration/mpi_job_controller_test.go [hunk 13]
```diff
@@ -822,6 +823,96 @@ func TestMPIJobWithVolcanoScheduler(t *testing.T) {
...
+		}
+		// 3. There should be no conditions, even the one for create
+		if mpiJobHasCondition(mpiJob, kubeflow.JobCreated) {
+			t.Errorf("MPIJob shouldn't have any condition")
+		}
+		// 4. No Pods or Services created
+		pods, err := getPodsForJob(ctx, s.kClient, mpiJob)
+		if err != nil {
+			t.Fatalf("Failed getting pods for the job: %v", err)
+		}
```
Comment (tenzen-y) [inline test/integration/mpi_job_controller_test.go:898]: Could we verify if the Launcher Job doesn't exist as well since the MPIOperator will create the Launcher batch/v1 Job and Worker Pods?

**Review by tenzen-y — COMMENTED** [review]
Thanks. I left minor comments.
Otherwise lgtm

File: test/e2e/mpi_job_test.go (hunk 12 above)
Comment (mszadkow) [reply test/e2e/mpi_job_test.go:192]: done

File: test/integration/mpi_job_controller_test.go (hunk 13 above)
Comment (mszadkow) [reply test/integration/mpi_job_controller_test.go:898]: done

File: pkg/apis/kubeflow/v2beta1/default.go [hunk 14]
```diff
@@ -53,6 +53,9 @@ func setDefaultsRunPolicy(policy *RunPolicy) {
 	if policy.CleanPodPolicy == nil {
 		policy.CleanPodPolicy = ptr.To(CleanPodPolicyNone)
 	}
+	if policy.ManagedBy == nil {
```
Comment (mimowo) [inline pkg/apis/kubeflow/v2beta1/default.go:56]: Why are we defaulting the field? It is a decision in other Job CRDs not to default it, so I would suggest to keep it this way, for consistency. It can also be helpful for MultiKueue. If we default the field here, then the MultiKueue webhook will need to override the value, rather than just set.

File: pkg/apis/kubeflow/v2beta1/types.go [hunk 15]
```diff
@@ -131,6 +139,16 @@ type RunPolicy struct {
...
+	// The mpi-operator reconciles a MPIJob which doesn't have this
+	// field at all or the field value is the reserved string
+	// 'kubeflow.org/mpi-operator', but delegates reconciling the MPIJob
+	// with 'kueue.x-k8s.io/multikueue' to the Kueue.
+	// The field is immutable.
```
Comment (mimowo) [inline pkg/apis/kubeflow/v2beta1/types.go:150]: Should add `// +optional` tag

File: test/integration/mpi_job_controller_test.go [hunk 16]
```diff
@@ -822,6 +823,96 @@ func TestMPIJobWithVolcanoScheduler(t *testing.T) {
...
+	mpiJob, err = s.mpiClient.KubeflowV2beta1().MPIJobs(s.namespace).Create(ctx, mpiJob, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Failed sending job to apiserver: %v", err)
+	}
+	backoff := wait.Backoff{
+		Duration: moderateTimeout,
+		Factor:   1.0,
+		Steps:    3,
+	}
+	err = wait.ExponentialBackoff(backoff, func() (bool, error) {
```
Comment (mimowo) [inline test/integration/mpi_job_controller_test.go:884]: What is the purpose of using `ExponentialBackoff` here? Can we just wait 1s before doing the asserts? I assume this will be enough to fail if ManagedBy wasn't suppored.

File: test/e2e/mpi_job_test.go [hunk 17]
```diff
@@ -164,8 +166,46 @@ var _ = ginkgo.Describe("MPIJob", func() {
...
 				expectConditionToBeTrue(mpiJob, kubeflow.JobSucceeded)
 			})
-		})
 
+			ginkgo.It("should not be updated when managed externaly, only created", func() {
+				mpiJob.Spec.RunPolicy.ManagedBy = ptr.To(kubeflow.MultiKueueController)
+				ctx := context.Background()
+				mpiJob = createJob(ctx, mpiJob)
+
+				gomega.Consistently(func() error {
```
Comment (mimowo) [inline test/e2e/mpi_job_test.go:175]: I don't think we need to retry the Get with Consistently. I think waiting 1s before the Get should be enough. If ManagedBy wasn't respected then the Get would return updated object.

File: test/integration/mpi_job_controller_test.go (hunk 16 above)
Comment (mszadkow) [reply test/integration/mpi_job_controller_test.go:884]: It's not very common to use plain sleep in tests, but sure it does the job

Comment (mimowo) [reply test/integration/mpi_job_controller_test.go]: By using `ExponentialBackoff` you do the sleep anyway, just hidden behind the scenes, and the reason is not clear. 

I think it is better to be transparent and comment on why we sleep, or name the var in a self-explanatory way. 

This is the approach we take in the core k8s: https://github.com/kubernetes/kubernetes/blob/0b91def6bcbdde82206ec779791762fb6bc337af/test/integration/job/job_test.go#L2108.

File: test/integration/mpi_job_controller_test.go [hunk 18]
```diff
@@ -876,41 +876,33 @@ func TestMPIJobManagedExternally(t *testing.T) {
...
-		}
-		if svcs != nil {
-			t.Fatalf("There should be no services from job: %v", svcs)
-		}
-		return false, nil
-	})
-	if !wait.Interrupted(err) {
-		t.Fatalf("Failed to verify externally managed mpiJob: %v", err)
+
+	time.Sleep(moderateTimeout)
```
Comment (mimowo) [inline test/integration/mpi_job_controller_test.go:880]: This is not really used as timeout here. Instead of using `moderateTimeout `use a variable which reflects the intention, sleepDurationForControllerLatency as in https://github.com/kubernetes/kubernetes/blob/master/test/integration/job/job_test.go#L2108 sgtm.

**Review by tenzen-y — COMMENTED** [review]
Thanks!
/approve

Leave lgtm on @alculquicondor

File: pkg/apis/kubeflow/v2beta1/types.go [hunk 19]
```diff
@@ -131,6 +139,17 @@ type RunPolicy struct {
 	// Defaults to false.
 	// +kubebuilder:default:=false
 	Suspend *bool `json:"suspend,omitempty"`
+
+	// ManagedBy is used to indicate the controller or entity that manages a MPIJob.
+	// The value must be either an empty, 'kubeflow.org/mpi-operator' or
```
Comment (alculquicondor) [inline pkg/apis/kubeflow/v2beta1/types.go:144]:
```suggestion
	// The value must be either empty, 'kubeflow.org/mpi-operator' or
```

**Review by alculquicondor — COMMENTED** [review]
/lgtm

File: pkg/apis/kubeflow/v2beta1/types.go (hunk 19 above)
Comment (tenzen-y) [reply pkg/apis/kubeflow/v2beta1/types.go:144]: @mszadkow Could you address this comment in a separate PR?

Comment (mszadkow) [reply pkg/apis/kubeflow/v2beta1/types.go:144]: Sure, will be addressed asap
